File: src-tauri/src/models/inventory.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub enum InventoryCategory {
    Runtime,
    WebServer,
    Database,
    BuildTool,
    Framework,
    PackageManager,
    DevTool,
}

impl InventoryCategory {
    pub fn all() -> Vec<InventoryCategory> {
        vec![
            InventoryCategory::Runtime,
            InventoryCategory::WebServer,
            InventoryCategory::Database,
            InventoryCategory::BuildTool,
            InventoryCategory::Framework,
            InventoryCategory::PackageManager,
            InventoryCategory::DevTool,
        ]
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub enum InstallSource {
    System,
    DevPort,
    Xampp,
    Laragon,
    Wamp,
    Scoop,
    Chocolatey,
    Manual,
    Unknown,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct InventoryItem {
    pub id: String,
    pub name: String,
    pub category: InventoryCategory,
    pub is_installed: bool,
    pub version: Option<String>,
    pub executable_path: Option<String>,
    pub install_source: InstallSource,
    pub is_running: bool,
    pub port: Option<u16>,
}

impl InventoryItem {
    pub fn new(id: String, name: String, category: InventoryCategory) -> Self {
        Self {
            id,
            name,
            category,
            is_installed: false,
            version: None,
            executable_path: None,
            install_source: InstallSource::Unknown,
            is_running: false,
            port: None,
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct InventoryResult {
    pub runtimes: Vec<InventoryItem>,
    pub web_servers: Vec<InventoryItem>,
    pub databases: Vec<InventoryItem>,
    pub build_tools: Vec<InventoryItem>,
    pub frameworks: Vec<InventoryItem>,
    pub package_managers: Vec<InventoryItem>,
    pub dev_tools: Vec<InventoryItem>,
    pub scanned_at: String,
    pub scan_duration_ms: u64,
}

impl InventoryResult {
    pub fn new() -> Self {
        Self {
            runtimes: Vec::new(),
            web_servers: Vec::new(),
            databases: Vec::new(),
            build_tools: Vec::new(),
            frameworks: Vec::new(),
            package_managers: Vec::new(),
            dev_tools: Vec::new(),
            scanned_at: String::new(),
            scan_duration_ms: 0,
        }
    }

    pub fn total_installed(&self) -> usize {
        self.runtimes.iter().filter(|i| i.is_installed).count()
            + self.web_servers.iter().filter(|i| i.is_installed).count()
            + self.databases.iter().filter(|i| i.is_installed).count()
            + self.build_tools.iter().filter(|i| i.is_installed).count()
            + self.frameworks.iter().filter(|i| i.is_installed).count()
            + self.package_managers.iter().filter(|i| i.is_installed).count()
            + self.dev_tools.iter().filter(|i| i.is_installed).count()
    }

    pub fn total_items(&self) -> usize {
        self.runtimes.len()
            + self.web_servers.len()
            + self.databases.len()
            + self.build_tools.len()
            + self.frameworks.len()
            + self.package_managers.len()
            + self.dev_tools.len()
    }
}

impl Default for InventoryResult {
    fn default() -> Self {
        Self::new()
    }
}

File: src/types/inventory.ts
import {
  Cpu,
  Globe,
  Database,
  Hammer,
  Layers,
  Package,
  Wrench,
  type LucideIcon,
} from "lucide-react";

export type InventoryCategory =
  | "runtime"
  | "webServer"
  | "database"
  | "buildTool"
  | "framework"
  | "packageManager"
  | "devTool";

export type InstallSource =
  | "system"
  | "devPort"
  | "xampp"
  | "laragon"
  | "wamp"
  | "scoop"
  | "chocolatey"
  | "manual"
  | "unknown";

export interface InventoryItem {
  id: string;
  name: string;
  category: InventoryCategory;
  isInstalled: boolean;
  version: string | null;
  executablePath: string | null;
  installSource: InstallSource;
  isRunning: boolean;
  port: number | null;
}

export interface InventoryResult {
  runtimes: InventoryItem[];
  webServers: InventoryItem[];
  databases: InventoryItem[];
  buildTools: InventoryItem[];
  frameworks: InventoryItem[];
  packageManagers: InventoryItem[];
  devTools: InventoryItem[];
  scannedAt: string;
  scanDurationMs: number;
}

export interface CategoryMeta {
  key: keyof Omit<InventoryResult, "scannedAt" | "scanDurationMs">;
  label: string;
  icon: LucideIcon;
  color: string;
}

export const CATEGORY_META: CategoryMeta[] = [
  {
    key: "webServers",
    label: "Web Servers",
    icon: Globe,
    color: "text-blue-400",
  },
  {
    key: "databases",
    label: "Databases",
    icon: Database,
    color: "text-green-400",
  },
  { key: "runtimes", label: "Runtimes", icon: Cpu, color: "text-yellow-400" },
  {
    key: "buildTools",
    label: "Build Tools",
    icon: Hammer,
    color: "text-orange-400",
  },
  {
    key: "frameworks",
    label: "Frameworks",
    icon: Layers,
    color: "text-purple-400",
  },
  {
    key: "packageManagers",
    label: "Package Managers",
    icon: Package,
    color: "text-cyan-400",
  },
  { key: "devTools", label: "Dev Tools", icon: Wrench, color: "text-pink-400" },
];

export const INSTALL_SOURCE_LABELS: Record<InstallSource, string> = {
  system: "System",
  devPort: "DevPort",
  xampp: "XAMPP",
  laragon: "Laragon",
  wamp: "WampServer",
  scoop: "Scoop",
  chocolatey: "Chocolatey",
  manual: "Manual",
  unknown: "Unknown",
};

export const INSTALL_SOURCE_COLORS: Record<InstallSource, string> = {
  system: "bg-slate-600",
  devPort: "bg-blue-600",
  xampp: "bg-orange-600",
  laragon: "bg-sky-600",
  wamp: "bg-pink-600",
  scoop: "bg-cyan-600",
  chocolatey: "bg-amber-600",
  manual: "bg-slate-500",
  unknown: "bg-slate-700",
};

// Inventory item ID → Service ID mapping
export const INVENTORY_TO_SERVICE_MAP: Record<string, string> = {
  apache: "apache",
  mariadb: "mariadb",
  mysql: "mariadb",
};

// Categories that can be controlled as services
export const CONTROLLABLE_CATEGORIES: InventoryCategory[] = ["webServer", "database"];
